### PR TITLE
Proposal: NetVC Context

### PR DESCRIPTION
**Pull request by oknet**
In the NetVConnection, we have get_local_addr() and get_remote_addr() method.
Also have members local_addr, remote_addr and netvc->con.addr.

Thus, we should using local_addr or remote_addr to replace member server_addr in UnixNetVConnection.
And SSLNetVConnection has member sslClientConnection and method setSSLClientConnection() and getSSLClientConnection() to indictor ATS is a client or server in a SSL session.

To abstract those two feature, I'm design the netvc context function.

As a proxy, there has two side: client side ( Client <-> Proxy ) and server side ( Proxy <-> Server ). With the netvc context funtion to indicate which side the NetVC working on.

### CODE REVIEW

**Comment by SolidWallOfCode** [comment]
I would like to avoid "client_addr" and "server_addr" as methods because those terms are frequently ambiguous. I would much prefer "local" and "remote" indicating the ATS address of the connection and the address of the other (remote) side of the connection.

**Comment by atsci** [comment]
Can one of the admins verify this patch? Only approve PRs which have been reviewed.

**Comment by zwoop** [comment]
Would it be possible to squash some of this commits down ? Or if not, can you assure that all intermediary stages are functional / compiles ?

**Comment by oknet** [comment]
@SolidWallOfCode Let me explain it. In the Client<-->ATS side, client_addr is Client IP Address and server_addr is ATS service IP Address. In the other side ( ATS <--> Server ), client_addr is ATS outgoing IP Address and server_addr is Remote Server IP Address. To working with NetVC Context, get_server_addr and get_client_addr are context-insensitive.

**Comment by oknet** [comment]
Currently, get_server_addr() and get_client_addr() implement as wrapper functions to call get_local_addr() and get_remote_addr() depend on netvc context. It is easy to replace server_addr member

**Comment by SolidWallOfCode** [comment]
My point would be the need to carefully explain what the methods do in different contexts is the problem. "Local" and "remote" are always clear and need little explanation and correspond much more to what is wanted - that is, the ATS address for the connection or the address of the other end of the connection. If get_server_addr() and get_client_addr() depend in context sensitive ways on get_local and get_remove, the simplest solution would seem to be to just remove get_server_addr() and get_client_addr() and directly call the underlying methods.

**Comment by oknet** [comment]
@SolidWallOfCode get_server_addr() and get_client_addr() is wrapper functions only. 

the wrapper function get_server_addr() is used to decide which variables should be used here.

for example ( iocore/net/SSLClientUtils.cc ):
```
 85     // Otherwise match by IP
 86     else {
 87       char buff[INET6_ADDRSTRLEN];                                                                                          
 88       ats_ip_ntop(netvc->server_addr, buff, INET6_ADDRSTRLEN);
 89       if (validate_hostname(cert, reinterpret_cast<unsigned char *>(buff), true, NULL)) {
 90         SSLDebug("IP %s verified OK", buff);
 91         return preverify_ok;
 92       }
 93       SSLDebug("IP verification failed for (%s)", buff);
 94     }
```

directly replace server_addr with get_local_addr() or get_remote_addr()
```
if (netvc->get_context() == NetVConnection_C2P) {
  ats_ip_ntop(netvc->get_local_addr(), buff, INET6_ADDRSTRLEN);
} else {
  ats_ip_ntop(netvc->get_remote_addr(), buff, INET6_ADDRSTRLEN);
}
```

replace server_addr with get_server_addr()
```
ats_ip_ntop(netvc->get_server_addr(), buff, INET6_ADDRSTRLEN);
```

**Comment by oknet** [comment]
review the code again, the variable name 'server_addr' confuse me. now, I'm using get_remote_addr() to replace server_addr and remote get_server_*() and get_client_*() methods. sorry about the mistaken.

**Comment by zwoop** [comment]
.

**Comment by zwoop** [comment]
@oknet There's now a merge conflict, can you rebase please?

**Comment by zwoop** [comment]
Assigned a number of people for review, this is a change with potentially dangerous implications. Also, @oknet, is it possible to squash this down to fewer commits?

**Comment by oknet** [comment]
@zwoop I will create another PR and rebase on the newest master branch.

**Comment by zwoop** [comment]
Updating this PR is fine too

File: iocore/net/UnixNetAccept.cc [hunk 1]
```diff
@@ -277,14 +277,12 @@ NetAccept::do_blocking_accept(EThread *t)
     vc->options.packet_mark = packet_mark;
     vc->options.packet_tos = packet_tos;
     vc->apply_options();
-    vc->from_accept_thread = true;
```
Comment (maskit) [inline iocore/net/UnixNetAccept.cc]: Should this line be removed within this change?
I'm fine with introducing the context, however, I guess it is not a part of it. This makes me nervous.

File: proxy/InkAPI.cc [hunk 2]
```diff
@@ -6249,12 +6249,12 @@ TSVConnFdCreate(int fd)
   vc->action_ = &a;
 
   vc->id = net_next_connection_number();
+  vc->set_context(Net_VConnection_P2S);
   vc->submit_time = Thread::get_hrtime();
   vc->set_is_transparent(false);
   vc->mutex = new_ProxyMutex();
 
   if (vc->connectUp(this_ethread(), fd) != CONNECT_SUCCESS) {
-    vc->free(this_ethread());
```
Comment (maskit) [inline proxy/InkAPI.cc]: This line too. I don't understand why we need to remove this line to use the context, though I'm not so familiar with these area.

Comment (oknet) [inline proxy/InkAPI.cc]: sorry, please refer to TS-4432, I file this to a dedicated jira issue.

File: iocore/net/UnixNetAccept.cc (hunk 1 above)
Comment (oknet) [inline iocore/net/UnixNetAccept.cc]: It's already set by getNetProcessor()->allocate_vc(NULL);

**Comment by oknet** [comment]
@zwoop a new PR#753 created based on the newest master branch.

**Comment by oknet** [comment]
@zwoop a new PR#753 created based on the newest master branch.

**Pull request closed without merging**